File: C++/787.cpp
__________________________________________________________________________________________________
sample 12 ms submission
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int K) {
        g_ = vector<vector<int>>(n, vector<int>(n, INT_MAX));
        for (int i = 0; i < n; ++i) g_[i][i] = 0;
        for (auto& f : flights) g_[f[0]][f[1]] = f[2];
        
        vector<int> mincost(n, INT_MAX);
        queue<pair<int, int>> q;  // node : cost
        
        K = K + 1; // to level
        q.push({src, 0});
        mincost[src] = 0;
        
        while (!q.empty() && K) {
            int size = q.size();
            
            for (int i = 0; i < size; ++i) {
                auto p = q.front(); q.pop();
                int now = p.first;
                int cost = p.second;
                for (int next = 0; next < n; ++next) {
                    if (g_[now][next] != INT_MAX && now != next) {
                        int cost_next = cost + g_[now][next];
                        if (cost_next >= mincost[next] || cost_next > mincost[dst]) continue;
                        q.push({next, cost_next});
                        mincost[next] = cost_next;
                    }
                }
            }
            --K;
        }
        return mincost[dst] == INT_MAX ? -1 : mincost[dst];
    }
    
    vector<vector<int>> g_;
};
__________________________________________________________________________________________________
sample 11156 kb submission
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int K) {
        vector<int> dis(n,1e8);
        dis[src]=0;
        for(int i=0;i<=K;i++){
            vector<int> x(dis);
            for(auto& j:flights)
                x[j[1]]=min(x[j[1]],dis[j[0]]+j[2]);
            dis=x;
        }
        return dis[dst]==1e8 ? -1 : dis[dst];
    }
};
__________________________________________________________________________________________________

File: Python3/787.py
__________________________________________________________________________________________________
sample 76 ms submission
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, K: int) -> int:
        
        ''' 
        #### BFS #### 
              
        graph = self.build_graph(n, flights)
        from collections import deque 
        queue = deque([(src, 0)])
        min_price = sys.maxsize 
        
        for k in range(K+1):
            
            visited = {}
            while queue:
                city, price = queue.popleft()
                    
                for next_city, next_price in graph[city]:
                    if next_city == dst:
                        min_price = min(min_price, price + next_price)
                        
                    if next_city in visited:
                        visited[next_city] = min(visited[next_city], price + next_price)
                    else:
                        visited[next_city] = price + next_price
                    
            queue = deque([(k, v) for k, v in visited.items()]) 
            
        return min_price if min_price < sys.maxsize else -1 
        '''
        
        #### Dijkstra's ####
        
        graph = self.build_graph(n, flights) 
        import heapq
        queue = [(0, src, -1)] #(price, city, stop)
        visited = set()
        while queue:
            cur_price, cur_city, stop = heapq.heappop(queue)
            if cur_city == dst:
                return cur_price 
            if stop >= K:
                continue   
            if (cur_city, stop) in visited:
                continue 
            visited.add((cur_city, stop))                
            for next_city, price in graph[cur_city]:
                heapq.heappush(queue, (price + cur_price, next_city, stop + 1))
                
        return -1 
       
    

    def build_graph(self, n, flights):
        graph = {i: [] for i in range(n)}
        for src, dst, price in flights:
            graph[src].append((dst, price))
        return graph 
__________________________________________________________________________________________________
sample 13624 kb submission
from typing import List
import math

class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, K: int) -> int:
        edges = dict(zip(range(n), (dict() for _ in range(n))))
        for start, end, dist in flights:
            edges[start][end] = dist
        
        l_prev = {src: 0}
        l_next = dict()
        min_dist = math.inf
        for _ in range(K + 1):
            for node in l_prev:
                for next_ in edges[node]:
                    cur_dist = l_prev[node] + edges[node][next_]
                    l_next[next_] = min(l_next[next_] if next_ in l_next else math.inf, cur_dist)
                    if next_ == dst:
                        min_dist = min(min_dist, l_next[next_])
            l_prev, l_next = l_next, dict()
        return -1 if min_dist == math.inf else min_dist
__________________________________________________________________________________________________
